### DISCUSSION TITLE
Add Privatevpn to VPN_PORT_FORWARDING_PROVIDER

### DISCUSSION THREAD

**Discussion by Silversurfer79** [qdm12/gluetun#2393] [Ideas]
Hi All,

Not sure if this the right forum, but is there anyway we can get priavetVPN added to the VPN_PORT_FORWARDING_PROVIDER list?

For what i know they support wireguard and port forward to dedicated servers, thanks.

**Comment by Silversurfer79** [comment]
Any reason there cannot be a generic option for other non supported VPN providers? I dont get there is a hard stop for this?

**Reply by qdm12** [reply]
> Any reason there cannot be a generic option for other non supported VPN providers?

Some providers assign you the forwarded port through their web ui, so you just need to set `FIREWALL_VPN_INPUT_PORTS`. That, funnily, was actually the case before for PrivateVPN.
Some other providers make it more complicated, such as ProtonVPN, Perfect Privacy, now PrivateVPN, and especially Private Internet Access (I hate them for it). In Privatevpn's case:

1. Establish the VPN
2. Find the assigned IP address of the tunnel interface
3. Send an http get request to `https://connect.pvdatanet.com/v3/Api/port?ip[]=yourassignedip`
4. Check the http status code
5. parse the json response
6. check the server supports port forwarding (boolean field)
7. extract the port number from the status json field

You could totally do this in a shell script, but isn't it more fun to just set `VPN_PORT_FORWARDING=on` (except the person having to create an issue)?? 😉  On that note, you could implement your own custom port forwarding logic once (I suggest you subscribe to both):
- Issue #1785 gets implemented (to trigger the port forwarding obtention command/program/script)
- PR #2392 gets merged, to allow you send the forwarded ports to the control server for Gluetun to 'adapt' the firewall. That would be the final step of your port forwarding script.

**Comment by qdm12** [comment]
> is there anyway we can get priavetVPN added to the VPN_PORT_FORWARDING_PROVIDER list?

See #1859
